pkg/config/store/fsstore2.go

```diff
@@ -0,0 +1,228 @@
+}
+
+// fsStore2 is Store2Backend implementation using filesystem.
+type fsStore2 struct {
+	root          string
+	kinds         map[string]bool
+	checkDuration time.Duration
+	watchCtx      context.Context
+	watchCh       chan BackendEvent
+
+	mu   sync.RWMutex
+	data map[Key]*resource
+}
+
+var _ Store2Backend = &fsStore2{}
+
+func parseFile(path string, data []byte) []*resource {
+	if bytes.HasPrefix(data, []byte("---\n")) {
+		data = data[4:]
+	}
+	if bytes.HasSuffix(data, []byte("\n")) {
+		data = data[:len(data)-1]
+	}
+	if bytes.HasSuffix(data, []byte("\n---")) {
+		data = data[:len(data)-4]
+	}
+	if len(data) == 0 {
+		return nil
+	}
+	chunks := bytes.Split(data, []byte("\n---\n"))
+	resources := make([]*resource, 0, len(chunks))
+	for i, chunk := range chunks {
+		r := &resource{}
+		if err := yaml.Unmarshal(chunk, r); err != nil {
+			glog.Errorf("Failed to parse %d-th part in file %s: %v", i, path, err)
+			continue
+		}
+		if r.Kind == "" || r.Metadata.Namespace == "" || r.Metadata.Name == "" {
+			glog.Errorf("Key elements are empty. Extracted as %s", r.Key())
+			continue
+		}
+		r.sha = sha1.Sum(chunk)
+		resources = append(resources, r)
+	}
+	return resources
+}
+
+func (s *fsStore2) readFiles() map[Key]*resource {
+	const suffix = ".yaml"
+	result := map[Key]*resource{}
+
+	err := filepath.Walk(s.root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !strings.HasSuffix(path, suffix) || (info.Mode()&os.ModeType) != 0 {
+			return nil
+		}
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			glog.Warningf("Failed to read %s: %v", path, err)
+			return err
+		}
+		for _, r := range parseFile(path, data) {
+			k := r.Key()
+			if !s.kinds[k.Kind] {
+				glog.Warningf("Unknown kind %s is in the file", k.Kind)
+				continue
+			}
+			result[r.Key()] = r
+		}
+		return nil
+	})
+	if err != nil {
+		glog.Errorf("failure during filepath.Walk: %v", err)
+	}
+	return result
+}
+
+func (s *fsStore2) update() {
+	newData := s.readFiles()
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	updated := []Key{}
+	removed := map[Key]bool{}
+	for k := range s.data {
+		removed[k] = true
+	}
+	for k, r := range newData {
+		oldR, ok := s.data[k]
+		if !ok {
+			updated = append(updated, k)
+			continue
+		}
+		delete(removed, k)
+		if r.sha != oldR.sha {
+			updated = append(updated, k)
+		}
+	}
+	if len(updated) == 0 && len(removed) == 0 {
+		return
+	}
+	s.data = newData
+	if s.watchCtx == nil || s.watchCtx.Err() != nil {
+		return
+	}
+	evs := make([]BackendEvent, 0, len(updated)+len(removed))
+	for _, key := range updated {
+		evs = append(evs, BackendEvent{Key: key, Type: Update, Value: s.data[key].Spec})
+	}
+	for key := range removed {
+		evs = append(evs, BackendEvent{Key: key, Type: Delete})
+	}
+	for _, ev := range evs {
+		select {
+		case <-s.watchCtx.Done():
+		case s.watchCh <- ev:
+		}
+	}
+}
+
+// NewFsStore2 creates a new Store2Backend backed by filesystem.
```

Comment:
nit: "backed by a filesystem", or "backed by the filesystem"

Reply:
done